awscli/customizations/assumerole.py

```diff
@@ -36,8 +38,41 @@ def inject_assume_role_provider_cache(session, **kwargs):
                   "assume-role cred provider cache.  Not configuring "
                   "JSONFileCache for assume-role.")
         return
-    provider = cred_chain.get_provider('assume-role')
-    provider.cache = JSONFileCache()
+    cred_chain.get_provider('assume-role').cache = JSONFileCache()
+    saml_provider = cred_chain.get_provider('assume-role-with-saml')
+    saml_provider.cache = JSONFileCache()
+    saml_provider.role_selector = role_selector
+
+
+def role_selector(role_arn, roles):
+    if role_arn:
+        chosen_roles = [r for r in roles if r['RoleArn']==role_arn]
+        if not chosen_roles:
+            raise ValueError(
+                'Your specified role %s does not exist in IdP roles list: %s'
+                % (role_arn, [r['RoleArn'] for r in roles]))
+        role = chosen_roles[0]
+    else:
+        # So end user has not yet configured a role_arn.
+        # If IdP provides more than one role, ask the user to choose one,
+        # otherwise just proceed.
+        if len(roles) > 1:
+            options = [(str(i), role) for i, role in enumerate(roles, start=1)]
+            prompt = "\n".join(
+                "[%s]: %s" % (i, role['RoleArn']) for i, role in options)
+            choice = raw_input(
+                "Please choose the role you would like to assume:\n"
+                "(You can preconfigure the role_arn in your profile)\n"
+                "%s\n"
+                "Selection: " % prompt)
+            role = dict(options).get(choice)
+            if not role:
+                raise ValueError('You selected an invalid role index.')
+        elif roles:
+            role = roles[0]
```

Comment:
I have to admit that the current "choosing role on the fly" and "automatically choose the only available role" behavior was influenced by [Quint's post](http://blogs.aws.amazon.com/security/post/TxU0AVUS9J00FP/How-to-Implement-a-General-Solution-for-Federated-API-CLI-Access-Using-SAML-2-0). He did the same thing there.

If I were an end user in an organization, I would just use whatever role my administrator gave me, be it "dev" or "marketing" or something else, as long as it allows me to finish the job that I was supposed to accomplish by that role. So, switching role is not a concern for me (as an end user). I can be an "devops" today and administrator gives me a role to handle EC2 instances, I switched to marketing department tomorrow and my new role can access materials in the organization's S3 bucket.

So do we want to have this feature here?